shared/src/store/action-reducers/utils/calculate-treatments.ts

```diff
@@ -1,255 +1,333 @@
-
-    switch (status) {
-        case 'red':
-            catersFor.red++;
-            break;
-        case 'yellow':
-            catersFor.yellow++;
-            break;
-        case 'green':
-            catersFor.green++;
-            break;
-        default:
-            break;
+
+    cateringElement.assignedPatientIds[patient.id] = true;
+
+    if (isPersonnel(cateringElement)) {
+        patient.assignedPersonnelIds[cateringElement.id] = true;
+        patient.isBeingTreated = true;
+    } else {
+        patient.assignedMaterialIds[cateringElement.id] = true;
     }
+
+    catersFor[status as 'green' | 'red' | 'yellow']++;
     return true;
 }
 
-export function calculateTreatments(state: Mutable<ExerciseState>) {
-    const pretriageEnabled = state.configuration.pretriageEnabled;
-    const bluePatientsEnabled = state.configuration.bluePatientsEnabled;
-    const personnels = Object.values(state.personnel).filter(
-        (personnel) => personnel.position !== undefined
-    );
-    const materials = Object.values(state.materials).filter(
-        (material) => material.position !== undefined
+// TODO: Instead, give each Element a "type" property -> discriminated union
+function isPatient(
+    element: Material | Patient | Personnel
+): element is Patient {
+    return (element as Patient).personalInformation !== undefined;
+}
+
+function isPersonnel(
+    element: Material | Patient | Personnel
+): element is Personnel {
+    return (element as Personnel).personnelType !== undefined;
+}
+
+function isMaterial(
+    element: Material | Patient | Personnel
+): element is Material {
+    // as Material does not include any distinguishable, we will check if it is not of type Personnel or Patient
+    return (
+        (element as Personnel).personnelType === undefined &&
+        (element as Patient).personalInformation === undefined
     );
-    // Unassign all patients as we are calculating everything from scratch.
-    personnels.forEach((personnel) => {
+}
+
+/**
+ * @param position of the patient where all elements of {@link elementType} should be recalculated
+ * @param elementIdsToBeSkipped the elements whose treatment should not be updated
+ */
+function updateCateringAroundPatient(
+    state: Mutable<ExerciseState>,
+    position: Position,
+    elementType: 'materials' | 'personnel',
+    elementIdsToBeSkipped: Set<UUID>
+) {
+    const elementsInTreatmentRange = SpatialTree.findAllElementsInCircle(
+        state.spatialTrees[elementType],
+        position,
+        maxTreatmentRange
+    ).filter((elementId) => !elementIdsToBeSkipped.has(elementId));
+
+    for (const elementId of elementsInTreatmentRange) {
+        updateCatering(state, getElement(state, elementType, elementId));
+    }
+}
+
+function removeTreatmentsOfElement(
+    state: Mutable<ExerciseState>,
+    element: Material | Patient | Personnel
+) {
+    if (isPatient(element)) {
+        const patient = getElement(state, 'patients', element.id);
```

Comment:
@ClFeSc What exactly do you mean?

Reply:
@Dassderdie @ClFeSc I tried removing the `getElement` and just write like `const patient = element;`.
This worked for `patient`, but not for `material` or for `personnel` (as the isXXX function does return not a `Mutable` type).
I could fix it but would be not good, I am also not sure why it works for `patient`.
Maybe writing the `isXXX` function to something that returns `Mutable` when it is really `Mutable` or something like that? Or we just keep it that way and maybe write `TODO` comment that this should be fixed when every element gets a type variable.

Reply:
> @Dassderdie @ClFeSc I tried removing the `getElement` and just write like `const patient = element;`. This worked for `patient`, but not for `material` or for `personnel` (as the `isXXX` functions do return not a `Mutable` type). I could fix it but would be not good (@Dassderdie maybe you know how to write this, like in `getElement`?), I am also not sure why it works for `patient`. Maybe writing the `isXXX` function to something that returns `Mutable` when it is really `Mutable` or something like that? Or we just keep it that way and maybe write `TODO` comment that this should be fixed when every element gets a type variable.

I don't know what you mean with the `isXXX` functions as they return a `boolean`, but you're completely right that this gets the mutable version of the `element` while `element` itself is immutable. So leave it like it is, my bad.

Reply:
> > @Dassderdie @ClFeSc I tried removing the `getElement` and just write like `const patient = element;`. This worked for `patient`, but not for `material` or for `personnel` (as the `isXXX` functions do return not a `Mutable` type). I could fix it but would be not good (@Dassderdie maybe you know how to write this, like in `getElement`?), I am also not sure why it works for `patient`. Maybe writing the `isXXX` function to something that returns `Mutable` when it is really `Mutable` or something like that? Or we just keep it that way and maybe write `TODO` comment that this should be fixed when every element gets a type variable.
> 
> I don't know what you mean with the `isXXX` functions as they return a `boolean`, but you're completely right that this gets the mutable version of the `element` while `element` itself is immutable. So leave it like it is, my bad.

I changed `element` in the input to be `Mutable` but typescript is clever (or not) enough to see that I check with `isPatient`, ... if the element is `Patient` or `Material` or `Personnel` (but not checking if `Mutable<Patient>` or `Mutable<Materiall>` and so on.
Even restarting TypeScript didn't help.
I don't get why it works for `patient` and not the rest, the `isXXX` are similiar enough that this shouldn't happen.

Should I change `element` in the input already to be `Mutable` and just write a comment like this?:
```
// TODO: when elements have their own type saved don't use const patient = getElement(state, 'patients', element.id);
// instead use const patient = element;
// same for personnel and material in the other if statements
```

`element` right now is always put into as `Mutable` anyway.

Reply:
Ah, I see. You could add the same as with `State` for `getElement` to these `isXXX` functions.

Reply:
That was an idea, do you know how to that (that it automatically results in `Mutable` when `Mutable` is entered and not when not?
Else, the `isXXX` functions could be changed to just accept `Mutable` (as it is only called with `Mutable` elements).

Reply:
As I said, you can do the same thing as with `State` in `getElement` (using a generic).

Reply:
> As I said, you can do the same thing as with `State` in `getElement` (using a generic).

I did this

```
function isPatient<PatientType extends Mutable<Patient> | Patient>(
    element: Material | PatientType | Personnel
): element is PatientType {
    return (element as PatientType).personalInformation !== undefined;
}

function isPersonnel<PersonnelType extends Mutable<Personnel> | Personnel>(
    element: Material | Patient | PersonnelType
): element is PersonnelType {
    return (element as PersonnelType).personnelType !== undefined;
}
```

For patient it results in element being after the `isPatient` being of `Mutable<Patient> | Mutable<Patient>` (yes twice).
and for personnel it results in `Personnel | Mutable<Personnel>` which does not work.

Reply:
I would just make these `isXXX` functions only take `Mutable` as this is what works right now, if somebody wants to use these functions with immutable elements before every element gets a type, this person can rewrite these functions.

Reply:
@ClFeSc are you ok with that or do you have a proposal to make it with generics work correctly?

Reply:
> ```ts
> function isPatient<PatientType extends Mutable<Patient> | Patient>(
>     element: Material | PatientType | Personnel
> ): element is PatientType {
>     return (element as PatientType).personalInformation !== undefined;
> }
> 
> function isPersonnel<PersonnelType extends Mutable<Personnel> | Personnel>(
>     element: Material | Patient | PersonnelType
> ): element is PersonnelType {
>     return (element as PersonnelType).personnelType !== undefined;
> }
> ```

I think typescript can't automatically infer the correct value for `PatientType` because the argument's type is a union. This should work by calling it like this `isPatient<Mutable<Patient>>(element)` or `isPatient<Patient>(element)`.

Also, keep in mind that this will soon be obsolete because of the `type` property in the element itself.

Reply:
After looking at your changes, I think you can leave it like this. 👍

Reply:
Yes, as this is obsolete when an element is changed, also making this compatible with immutable elements, when right now it is called everytime with `Mutable` would make it more complicated than it needs to be (KISS)